Kernel/Assertions.cpp

```diff
@@ -35,8 +29,18 @@ void __assertion_failed(char const* msg, char const* file, unsigned line, char c
     PANIC("Aborted");
 }
 
+void __assertion_failed(char const* msg, char const* file, unsigned line, char const* func)
+{
+    Processor::disable_interrupts();
+
+    critical_dmesgln("ASSERTION FAILED: {}", msg);
+    critical_dmesgln("{}:{} in {}", file, line, func);
+
+    abort();
+}
+
 [[noreturn]] void _abort()
 {
-    asm volatile("ud2");
+    Processor::undefined_behaviour();
```

Comment:
Did you actually check what happens when we hit an assertion failure now? Because I think this instruction will generate a synchronous exception, which will call `exception_common`, which will choke on this exception since it is an unexpected exception, and then call `PANIC` again, ending up in an infinite loop..

Reply:
nope; analogous instruction (triggering an exception too) is used in x86-64 code and did go blindly here—iʼd still abstract that.

would you accept making `undefined_behaviour` a to do for Aarch64 and comment why we donʼt want to have it there yet? (`__builtin_unreachable` can have similar effect, though)

Reply:
> nope; analogous instruction (triggering an exception too) is used in x86-64 code and did go blindly here—iʼd still abstract that.

Yes, but the x86-64 code actually handles the exception (I think somewhere in Arch/x86_64/Interrupts.cpp). The current aarch64 does not handle the `udf 0` exception, which will result in an infinite loop of exceptions, which is why you should actually test this code.

Also, I think `Processor::undefined_behaviour()` is kinda a bad name, because this will cause totally normal (defined) behaviour to be executed (which in this case is crashing the kernel).